NetCasbin/Rbac/DefaultRoleManager.cs

```diff
@@ -1,163 +1,271 @@
+                : HasLink(name1, name2, domain[0]);
+        }
 
-        public void AddLink(string name1, string name2, params string[] domain)
+        public virtual void AddLink(string name1, string name2, params string[] domain)
         {
-            if (domain.Length == 1)
+            CheckDomainArgument(domain);
+            if (domain.Length is 0)
             {
-                name1 = $"{domain[0]}::{name1}";
-                name2 = $"{domain[0]}::{name2}";
+                AddLink(name1, name2);
             }
-            else if (domain.Length > 1)
+            else
             {
-                throw new ArgumentException(_domainError);
+                AddLink(name1, name2, domain[0]);
             }
-
-            var role1 = CreateRole(name1);
-            var role2 = CreateRole(name2);
-            role1.AddRole(role2);
-        }
-
-        public virtual void Clear()
-        {
-            _allRoles.Clear();
         }
 
         public virtual void DeleteLink(string name1, string name2, params string[] domain)
         {
-            if (domain.Length == 1)
+            CheckDomainArgument(domain);
+            if (domain.Length is 0)
             {
-                name1 = PrefixDomain(domain[0], name1);
-                name2 = PrefixDomain(domain[0], name2);
+                DeleteLink(name1, name2);
             }
-            else if (domain.Length > 1)
+            else
             {
-                throw new ArgumentException(_domainError);
+                DeleteLink(name1, name2, domain[0]);
             }
+        }
 
-            if (!HasRole(name1) || !HasRole(name2))
+        private void CheckDomainArgument(string[] domain)
+        {
+            if (domain.Length > 1)
             {
-                return;
+                throw new ArgumentException(" Domain should be 1 parameter.");
             }
-
-            var role1 = CreateRole(name1);
-            var role2 = CreateRole(name2);
-            role1.DeleteRole(role2);
         }
+        #endregion
 
-        public virtual List<string> GetRoles(string name, params string[] domain)
+        private IEnumerable<string> GetRoles(string name, string domain = null)
         {
-            if (domain.Length == 1)
+            domain ??= _defaultDomain;
+            ConcurrentDictionary<string, Role> roles;
+            if (HasPattern || HasDomainPattern)
             {
-                name = domain[0] + "::" + name;
+                roles = GenerateTempRoles(domain);
             }
-            else if (domain.Length > 1)
+            else
             {
-                throw new Exception("error: domain should be 1 parameter");
+                if (_allDomains.TryGetValue(domain, out roles) is false)
+                {
+                    return Enumerable.Empty<string>();
+                }
             }
 
-            if (!HasRole(name))
+            Func<string, string, bool> matchingFunc = MatchingFunc;
```

Comment:
No need to copy another `matchingFunc`, you can use `MatchingFunc` directly

Reply:
I know, but it only likes a pointer and there is almost no difference between the two writing methods. I just do not want to call the get method of MatchingFunc twice.

Reply:
I don't know the difference using `matchingFunc` and `MatchingFunc`, but it's Ok for me.